### PR TITLE
Feature/SOF-6588-1 feature: working exaxyz linter

### PR DESCRIPTION
**Pull request by VsevolodX**
This is branching off the previous branch, containing the latest advances in tackling the issue

### CODE REVIEW

File: src/other/codemirror/CodeMirror.tsx [hunk 1]
```diff
@@ -30,69 +33,111 @@ export interface CodeMirrorProps {
...
+            isLoaded: false,
+            isEditing: false,
+            extensions: this.createExtensions(),
+        };
         this.handleContentChange = this.handleContentChange.bind(this);
+        this.handleFocus = this.handleFocus.bind(this);
+        this.handleBlur = this.handleBlur.bind(this);
+    }
+
+    componentDidUpdate(prevProps: CodeMirrorProps) {
```
Comment (timurbazhirov) [inline src/other/codemirror/CodeMirror.tsx]: componentWillReceiveProps

File: src/other/codemirror/CodeMirror.tsx [hunk 2]
```diff
@@ -30,69 +33,111 @@ export interface CodeMirrorProps {
...
+            if (isEditing && updateContent) updateContent(newContent);
+        });
     }
 
-    // eslint-disable-next-line class-methods-use-this
-    getLanguageExtensions(language: string) {
-        if (LANGUAGES_MAP[language]) return LANGUAGES_MAP[language];
+    handleFocus() {
+        const { onFocus } = this.props;
+        if (onFocus) onFocus();
```
Comment (timurbazhirov) [inline src/other/codemirror/CodeMirror.tsx]: We should not pass these onFocus, onBlur to this component

File: src/other/codemirror/CodeMirror.tsx [hunk 3]
```diff
@@ -30,69 +33,111 @@ export interface CodeMirrorProps {
...
+            const linterExtension = linterGenerator(checks);
+            return [completionExtension, linter(linterExtension), ...languageExtensions];
+        }
+
+        return [completionExtension, ...languageExtensions];
+    }
+
+    render() {
+        const { options = {}, theme, readOnly, content, triggerReload } = this.props;
+        const { extensions } = this.state;
```
Comment (timurbazhirov) [inline src/other/codemirror/CodeMirror.tsx]: so is content

File: src/other/codemirror/utils/exaxyz_linter.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+import _ from "underscore";
+
+const linterGenerator = (checks: ConsistencyCheck[]) => {
+    return (view: EditorView): Diagnostic[] => {
+        const { doc } = view.state;
+        if (!checks && !doc) return [];
+
+        const warnings = checks.map((check) => {
+            const keyFragments = check.key.split(".");
+            const atomIdStr = _.last(keyFragments);
```
Comment (timurbazhirov) [inline src/other/codemirror/utils/exaxyz_linter.ts]: AtomId is a Material domain - it should not be in Cove, which is generic.

linterGenerator, in principle, is a generic enough to be here

File: src/other/codemirror/CodeMirror.tsx [hunk 5]
```diff
@@ -30,69 +33,111 @@ export interface CodeMirrorProps {
...
-        const { content = "", options = {}, language, completions } = this.props;
+    createExtensions(checks?: ConsistencyCheck[]): Extension[] {
+        const { completions, language } = this.props;
         const completionExtension = autocompletion({ override: [completions] });
+        const languageExtensions = LANGUAGES_MAP[language]
+            ? LANGUAGES_MAP[language]
+            : LANGUAGES_MAP.fortran;
+
+        if (checks) {
+            const linterExtension = linterGenerator(checks);
```
Comment (timurbazhirov) [inline src/other/codemirror/CodeMirror.tsx:89]: checks should be state variable

File: src/other/codemirror/utils/exaxyz_linter.ts (hunk 4 above)
Comment (VsevolodX) [reply src/other/codemirror/utils/exaxyz_linter.ts]: Yes. Basically, it's just the difference in naming. The logic itself is generic already. renaming: AtomId -> key

File: src/other/codemirror/CodeMirror.tsx [hunk 6]
```diff
@@ -22,77 +25,93 @@ const LANGUAGES_MAP: Record<string, any> = {
...
-            this.setState({ isLoaded: true });
-            return;
+    UNSAFE_componentWillReceiveProps(nextProps: CodeMirrorProps) {
+        const { content, checks } = this.props;
+        if (nextProps.content !== content) {
+            this.setState({ content: nextProps.content || "" });
+        }
+
+        if (nextProps.checks !== checks) {
+            this.setState({ checks: nextProps.checks });
```
Comment (timurbazhirov) [inline src/other/codemirror/CodeMirror.tsx]: Probably better to have the same condition as on line 73 and one `setState` accordingly

**Review by seankwarren — APPROVED** [review]

File: src/other/codemirror/CodeMirror.tsx (hunk 6 above)
Comment (VsevolodX) [reply src/other/codemirror/CodeMirror.tsx]: Nope. This was a mistake to change. The logic that was here is correct and for a reason. Content and checks must be updated independetly

**Pull request merged**